File: src/main/java/net/ndrei/teslacorelib/gui/TeslaEnergyLevelPiece.java
package net.ndrei.teslacorelib.gui;

import com.google.common.collect.Lists;
import com.mojang.realmsclient.gui.ChatFormatting;
import net.ndrei.teslacorelib.inventory.EnergyStorage;

import java.util.List;

/**
 * Created by CF on 2016-12-22.
 */
public class TeslaEnergyLevelPiece extends BasicContainerGuiPiece {
    private EnergyStorage energyStorage;

    public TeslaEnergyLevelPiece(int left, int top, EnergyStorage energyStorage) {
        super(left, top, 18, 54);
        this.energyStorage = energyStorage;
    }

    @Override
    public void drawBackgroundLayer(BasicTeslaGuiContainer container, int guiX, int guiY, float partialTicks, int mouseX, int mouseY) {
        BasicTeslaGuiContainer.bindDefaultTexture(container);

        container.drawTexturedRect(this.getLeft(), this.getTop(),
                1, 189, this.getWidth(), this.getHeight());

        /*if (Loader.isModLoaded("tesla")) {
            this.renderTeslaBackground(container, guiX, guiY);
        }
        else*/ if (this.energyStorage != null) {
            int power = (this.energyStorage.getEnergyStored() * (this.getHeight() - 6)) / this.energyStorage.getMaxEnergyStored();

            container.drawTexturedRect(this.getLeft() + 2, this.getTop() + 2, 20, 191, this.getWidth() - 4, this.getHeight() - 4);
            container.drawTexturedRect(this.getLeft() + 3, this.getTop() + 3 + this.getHeight() - 6 - power, 35, 192 + this.getHeight() - 6 - power, this.getWidth() - 6, power + 2);
        }
    }

//    @Optional.Method(modid = "tesla")
//    private void renderTeslaBackground(BasicTeslaGuiContainer container, int guiX, int guiY) {
//        if (this.energyStorage != null) {
//            PowerBar bar = new PowerBar(container,
//                    guiX + super.getLeft() + 2,  guiY + this.getTop() + 2,
//                    PowerBar.BackgroundType.LIGHT);
//            bar.draw(this.energyStorage);
//        }
//    }

    @Override
    public void drawForegroundTopLayer(BasicTeslaGuiContainer container, int guiX, int guiY, int mouseX, int mouseY) {
        if (super.isInside(container, mouseX, mouseY) && (this.energyStorage != null)) {
            List<String> lines = Lists.newArrayList();
            lines.add(String.format("%sStored Energy", ChatFormatting.DARK_PURPLE));
            lines.add(String.format("%s%,d T %sof", ChatFormatting.AQUA, this.energyStorage.getEnergyStored(), ChatFormatting.DARK_GRAY));
            lines.add(String.format("%s%,d T", ChatFormatting.RESET, this.energyStorage.getMaxEnergyStored()));

            long tick = this.energyStorage.getLastTickEnergy(), average = this.energyStorage.getAverageEnergyPerTick();
            lines.add(String.format("%s%,d T %s(%s%,d T%s)",
                    (average < 0) ? ChatFormatting.RED : ChatFormatting.BLUE, average,
                    ChatFormatting.RESET,
                    (tick < 0) ? ChatFormatting.RED : ChatFormatting.BLUE, tick,
                    ChatFormatting.RESET
            ));
            container.drawTooltip(lines, mouseX - guiX, mouseY - guiY);
        }
    }
}

File: src/main/kotlin/net/ndrei/teslacorelib/gui/TeslaEnergyLevelPiece.kt
package net.ndrei.teslacorelib.gui

import net.minecraft.util.text.TextFormatting
import net.ndrei.teslacorelib.MOD_ID
import net.ndrei.teslacorelib.config.TeslaCoreLibConfig
import net.ndrei.teslacorelib.inventory.EnergyStorage
import net.ndrei.teslacorelib.localization.GUI_ENERGY
import net.ndrei.teslacorelib.localization.localizeModString

/**
 * Created by CF on 2017-06-28.
 */
class TeslaEnergyLevelPiece(left: Int, top: Int, private val energyStorage: EnergyStorage?) : BasicContainerGuiPiece(left, top, 18, 54) {
    private val displayType get() = TeslaCoreLibConfig.energyDisplay

    override fun drawBackgroundLayer(container: BasicTeslaGuiContainer<*>, guiX: Int, guiY: Int, partialTicks: Float, mouseX: Int, mouseY: Int) {
        BasicTeslaGuiContainer.bindDefaultTexture(container)

        container.drawTexturedRect(this.left, this.top, 1, 189, this.width, this.height)
        container.drawTexturedRect(this.left + 2, this.top + 2, 20, 191, this.width - 4, this.height - 4)

        if (this.energyStorage != null) {
            val stored = this.displayType.fromTesla(this.energyStorage.stored)
            val capacity = this.displayType.fromTesla(this.energyStorage.capacity)
            val power = (stored * this.displayType.fullIcon.height / capacity).toInt()

            container.drawTexturedRect(this.left + 3, this.top + 3,
                //20, 191,
                this.displayType.emptyIcon.left, this.displayType.emptyIcon.top,
                this.displayType.emptyIcon.width, this.displayType.emptyIcon.height)
            container.drawTexturedRect(this.left + 3, this.top + 3 + this.displayType.fullIcon.height - power,
                this.displayType.fullIcon.left, this.displayType.fullIcon.top + this.displayType.fullIcon.height - power,
                this.displayType.fullIcon.width, power)
        }
    }

    override fun drawForegroundTopLayer(container: BasicTeslaGuiContainer<*>, guiX: Int, guiY: Int, mouseX: Int, mouseY: Int) {
        if (super.isInside(container, mouseX, mouseY) && (this.energyStorage != null)) {
            val tick = this.energyStorage.lastTickEnergy
            val average = this.energyStorage.averageEnergyPerTick

            val lines = mutableListOf(
                localizeModString(MOD_ID, GUI_ENERGY, "stored energy") {
                    +TextFormatting.DARK_PURPLE
                },
                localizeModString(MOD_ID, GUI_ENERGY, "stored line 1") {
                    +TextFormatting.GRAY
                    +this@TeslaEnergyLevelPiece.displayType.makeLightTextComponent(this@TeslaEnergyLevelPiece.energyStorage.stored)
                },
                localizeModString(MOD_ID, GUI_ENERGY, "stored line 2") {
                    +TextFormatting.GRAY
                    +this@TeslaEnergyLevelPiece.displayType.makeDarkTextComponent(this@TeslaEnergyLevelPiece.energyStorage.capacity)
                },
                localizeModString(MOD_ID, GUI_ENERGY, "statistic") {
                    +this@TeslaEnergyLevelPiece.displayType.makeTextComponent(average, if (average < 0) TextFormatting.RED else TextFormatting.BLUE)
                    +this@TeslaEnergyLevelPiece.displayType.makeTextComponent(tick, if (tick < 0) TextFormatting.RED else TextFormatting.BLUE)
                }
            ).map { it.formattedText }
            container.drawTooltip(lines, mouseX - guiX, mouseY - guiY)
        }
    }

    override fun mouseClicked(container: BasicTeslaGuiContainer<*>, mouseX: Int, mouseY: Int, mouseButton: Int) {
        super.mouseClicked(container, mouseX, mouseY, mouseButton)

        if (this.isInside(container, mouseX, mouseY) && TeslaCoreLibConfig.getFlag(TeslaCoreLibConfig.ALLOW_ENERGY_DISPLAY_CHANGE)) {
            TeslaCoreLibConfig.energyDisplay = EnergyDisplayType.values()[
                (TeslaCoreLibConfig.energyDisplay.ordinal + 1) % EnergyDisplayType.values().size
                ]
        }
    }
}
